pkg/instancegroups/instancegroups.go

```diff
@@ -290,6 +300,84 @@ func (r *RollingUpdateInstanceGroup) patchTaint(rollingUpdateData *RollingUpdate
 	return err
 }
 
+func (r *RollingUpdateInstanceGroup) drainTerminateAndWait(u *cloudinstances.CloudInstanceGroupMember, rollingUpdateData *RollingUpdateCluster, terminateChan chan error, isBastion bool, sleepAfterTerminate time.Duration) {
+	instanceId := u.ID
+
+	nodeName := ""
+	if u.Node != nil {
+		nodeName = u.Node.Name
+	}
+
+	if isBastion {
+		// We don't want to validate for bastions - they aren't part of the cluster
+	} else if rollingUpdateData.CloudOnly {
+
+		klog.Warning("Not draining cluster nodes as 'cloudonly' flag is set.")
+
+	} else {
+
+		if u.Node != nil {
+			klog.Infof("Draining the node: %q.", nodeName)
+
+			if err := r.DrainNode(u, rollingUpdateData); err != nil {
+				if rollingUpdateData.FailOnDrainError {
+					terminateChan <- fmt.Errorf("failed to drain node %q: %v", nodeName, err)
+					return
+				}
+				klog.Infof("Ignoring error draining node %q: %v", nodeName, err)
+			}
+		} else {
+			klog.Warningf("Skipping drain of instance %q, because it is not registered in kubernetes", instanceId)
+		}
+	}
+
+	// We unregister the node before deleting it; if the replacement comes up with the same name it would otherwise still be cordoned
+	// (It often seems like GCE tries to re-use names)
+	if !isBastion && !rollingUpdateData.CloudOnly {
+		if u.Node == nil {
+			klog.Warningf("no kubernetes Node associated with %s, skipping node deletion", instanceId)
+		} else {
+			klog.Infof("deleting node %q from kubernetes", nodeName)
+			if err := r.deleteNode(u.Node, rollingUpdateData); err != nil {
+				terminateChan <- fmt.Errorf("error deleting node %q: %v", nodeName, err)
+				return
+			}
+		}
+	}
+
+	if err := r.DeleteInstance(u); err != nil {
+		klog.Errorf("error deleting instance %q, node %q: %v", instanceId, nodeName, err)
+		terminateChan <- err
+		return
+	}
+
+	// Wait for the minimum interval
+	klog.Infof("waiting for %v after terminating instance", sleepAfterTerminate)
+	time.Sleep(sleepAfterTerminate)
+
+	terminateChan <- nil
```

Comment:
It's more composable and less error-prone not to pass the channel in, IMO; but rather to return the error, and then call it with something like  `go result <- foo()` .  Then you can call it without needing channels; you know in the caller that the caller that the error is always sent to the channel (i.e. the parallelism logic is in one function) and e.g. you don't risk pushing a value onto the channel and forgetting to return.

Reply:
Done